### PR TITLE
Code generator

### PR DESCRIPTION
**Pull request by S3JER**

### CODE REVIEW

File: GOAT-Compiler/Code Generation/CNCMachine.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        private double _extruderTemp = 0;
+        private double _fanPower = 0;
+
+        /// <summary>
+        /// The current positon of the extruder.
+        /// </summary>
+        public Vector Position { get; set; } = new Vector(0, 0, 0);
+
+        /// <summary>
+        /// The current extrusion.
```
Comment (henneboy) [inline GOAT-Compiler/Code Generation/CNCMachine.cs:30]: Plz give better summary, this is the current amount of extrusion, aka. distance*extrusionRate ?

File: GOAT-Compiler/Code Generation/CNCMachine.cs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            } 
+        }
+
+        /// <summary>
+        /// The current extruder rate.
+        /// </summary>
+        public double ExtrusionRate { get; set; } = 0;
+        
+        /// <summary>
+        /// The current bed temperature.
```
Comment (henneboy) [inline GOAT-Compiler/Code Generation/CNCMachine.cs:54]: Plz define which bed? Aka the hot-bed of the printer

File: GOAT-Compiler/RecSymbolTable.cs [hunk 3]
```diff
@@ -91,6 +93,11 @@ public Node GetFunctionNode(Symbol symbol)
         {
             return funcDeclMap.TryGetValue(symbol, out Node node) ? node : null;
         }
+
+        public bool IsGlobal(Symbol symbol)
```
Comment (henneboy) [inline GOAT-Compiler/RecSymbolTable.cs:97]: plz turn into arrow function

File: VisitorTests/CodeGenerator/BuiltInFunctionsTests.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            // do the parsing and checking needed
+            Start s = FileReadingTestUtilities.ParseString(GOATCode);
+            ISymbolTable symbolTable = FileReadingTestUtilities.BuildSymbolTable(s);
+            TypeChecker typeChecker = FileReadingTestUtilities.DoTypeChecking(s, symbolTable);
+            
+            // run the code generator
+            StringBuilder stringBuilder = new StringBuilder();
+            using (StringWriter writer = new StringWriter(stringBuilder))
+            {
+                CodeGenerator codeGenerator;
```
Comment (henneboy) [inline VisitorTests/CodeGenerator/BuiltInFunctionsTests.cs:37]: Plz move the declaration down to the next line, this line is redundant :)

Comment (henneboy) [reply VisitorTests/CodeGenerator/BuiltInFunctionsTests.cs:37]: Aka: remove this line and add "CodeGenerator" as prefix for the next line

File: VisitorTests/CodeGenerator/CodeGeneratorTest.cs [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            using (Stream stream = new FileStream(outPutFileName, FileMode.Open))
+            {
+                using (StreamWriter writer = new StreamWriter(stream))
+                {
+                    codeGenerator = new CodeGenerator(symbolTable, typeChecker.GetTypeDictionary(), writer);
+                    s.Apply(codeGenerator);
+                }
+            }
+
+            Assert.True(File.Exists(outPutFileName));
```
Comment (henneboy) [inline VisitorTests/CodeGenerator/CodeGeneratorTest.cs:38]: Assert followed by and if? Something is wrong here

File: VisitorTests/CodeGenerator/CodeGeneratorTest.cs [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        [SkippableTheory(typeof(TestDependencyException))]
+        [InlineData("RelMoveTest.txt", "RelMoveTest.gcode")]
+        public void CreatedFileChecker(string file, string outPutFileName)
+        {
+            string TestFilePath = FileReadingTestUtilities.ProjectBaseDirectory + "CodeGenerator/Tests/InputFolder/"+ file;
+            Start s = FileReadingTestUtilities.ParseFile(TestFilePath);
+            ISymbolTable symbolTable = FileReadingTestUtilities.BuildSymbolTable(s);
+            TypeChecker typeChecker = FileReadingTestUtilities.DoTypeChecking(s, symbolTable);
+            
+            CodeGenerator codeGenerator;
```
Comment (henneboy) [inline VisitorTests/CodeGenerator/CodeGeneratorTest.cs:26]: Dont declare codeGenerator here, just add the type to the line where it's used

File: VisitorTests/CodeGenerator/CodeGeneratorTest.cs [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+                using (StreamWriter writer = new StreamWriter(stream))
+                {
+                    codeGenerator = new CodeGenerator(symbolTable, typeChecker.GetTypeDictionary(), writer);
+                    s.Apply(codeGenerator);
+                }
+            }
+            Assert.True(File.Exists(outPutFileName));
+            
+            Assert.True(File.ReadAllText(outPutFileName).Length > 0);
+            if (File.Exists(outPutFileName))
```
Comment (henneboy) [inline VisitorTests/CodeGenerator/CodeGeneratorTest.cs:67]: redundant if?

File: VisitorTests/CodeGenerator/CodeGeneratorTest.cs [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+            using (MemoryStream dummyStream = new MemoryStream())
+            {
+                using (StreamWriter writer = new StreamWriter(dummyStream))
+                {
+                    generator = new CodeGenerator(symbolTable, typeChecker.GetTypeDictionary(), writer);
+                    s.Apply(generator);
+                }
+            }
+
```
Comment (henneboy) [inline VisitorTests/CodeGenerator/CodeGeneratorTest.cs:103]: too many empty lines

File: GOAT-Compiler/TypeChecker.cs [hunk 9]
```diff
@@ -15,22 +15,27 @@ internal class TypeChecker : SymbolTableVisitor
     {
         private Types currentFunctionType;
         private Dictionary<Node, Types> _typeDictionary = new Dictionary<Node, Types>();
-        
+
+        public Dictionary<Node, Types> GetTypeDictionary()
```
Comment (henneboy) [inline GOAT-Compiler/TypeChecker.cs:19]: arrowfunction

File: GOAT-Compiler/TypeChecker.cs [hunk 10]
```diff
@@ -414,7 +428,7 @@ public override void OutAReturnStmt(AReturnStmt node)
         {
             if (currentFunctionType != Convert(node.GetExp(), currentFunctionType))
             {
-                throw new TypeMismatchException(node);
+                throw new TypeMismatchException(node, "Wrong type of return statement");
             }
             else
```
Comment (henneboy) [inline GOAT-Compiler/TypeChecker.cs:433]: remove this else

File: GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+            _stream = stream;
+        }
+
+        /// <summary>
+        /// Calls the specified built in function with the right params.
+        /// </summary>
+        /// <param name="functionName">The name of the built in function.</param>
+        /// <param name="actuelParams">The params for the built in function.</param>
+        /// <returns></returns>
+        public dynamic CallBuildInFunctions(string functionName, List<dynamic> actuelParams)
```
Comment (henneboy) [inline GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs:30]: actuelParams is writen veru got, *actualParams

File: GOAT-Compiler/Code Generation/CNCMachine.cs [hunk 12]
```diff
@@ -0,0 +1,121 @@
+﻿using System;
+
+namespace GOAT_Compiler.Code_Generation
+{
+
+    public enum BuildScope
```
Comment (henneboy) [inline GOAT-Compiler/Code Generation/CNCMachine.cs:6]: nonono, plz rename to ExtrusionMode

File: GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+            string gLine = "G28";
+            _machine.Position = new Vector(0, 0, 0);
+            _stream.WriteLine(gLine);
+        }
+
+        private void RelMove(Vector v)
+        {
+            string gLine = "";
+            Vector oldPosition = _machine.Position;
+            _machine.Position = new Vector(oldPosition.X + v.X, oldPosition.Y + v.Y, oldPosition.Z + v.Z);
```
Comment (henneboy) [inline GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs:113]: why not just _machine.Position +=v?

File: GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+            if (_machine.Build == BuildScope.build)
+            {
+                _machine.CurrentExtrusion += (_machine.ExtrusionRate*VectorDistance(oldPosition, _machine.Position));
+                gLine += "G1 X" + (decimal)_machine.Position.X + " Y" + (decimal)_machine.Position.Y + " Z" + (decimal)_machine.Position.Z + " E" + (decimal)_machine.CurrentExtrusion;
```
Comment (henneboy) [inline GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs:117]: plz make a helper function that converts a vector to a gcode position:
string f-name(Vector v)=> $"X{(decimal)v.x} Y{(decimal)v.y} Z{(decimal)v.z}";

File: GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+            }
+            else if (_machine.Build == BuildScope.walk)
+            {
+                gLine += "G0 X" + (decimal)_machine.Position.X + " Y" + (decimal)_machine.Position.Y + " Z" + (decimal)_machine.Position.Z;
+            }
+            else
```
Comment (henneboy) [inline GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs:123]: plz make helper function which takes the _michine.Build and a string of the name of the command, and throws an exception if necessary:
void f(BuildScope machineExtrusionMode, string MovementFunctionName){
if (machineExtrusionMode == BuildScope.none)
 MoveWithoutScopeException($"{MovementFunctionName} cant be called without being in a build or walk scope.");
}

File: GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs [hunk 16]
```diff
@@ -0,0 +1,343 @@
...
+            _machine.ExtrusionRate = rate;
+        }
+        private void SetBedTemp(double temp)
+        {
+            string gLine = "";
+            _machine.BedTemp = temp;
+            gLine = "M140 S" + (decimal)temp;
+            _stream.WriteLine(gLine);
+        }
+        private Vector Position(){
```
Comment (henneboy) [inline GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs:273]: arrowfunction

File: GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs [hunk 17]
```diff
@@ -0,0 +1,343 @@
...
+        }
+        private void SetFanPower(double power)
+        {
+            string gLine = "";
+            _machine.FanPower = power;
+            gLine = "M106 S" + Math.Floor(_machine.FanPower * 255);
+            _stream.WriteLine(gLine);
+
+        }
+        private void SetExtrusionRate(double rate)
```
Comment (henneboy) [inline GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs:262]: arrowfunction

File: GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs [hunk 18]
```diff
@@ -0,0 +1,343 @@
...
+            {
+                gLine += "G0 X" + (decimal)_machine.Position.X + " Y" + (decimal)_machine.Position.Y + " Z" + (decimal)_machine.Position.Z;
+            }
+            else
+            {
+                throw new MoveWithoutScopeException("Lift cant be called without being in a build or walk scope.");
+            }
+            _stream.WriteLine(gLine);
+        }
+        private void Right(double deg)
```
Comment (henneboy) [inline GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs:306]: arrowfunction for all of these!

File: GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs [hunk 19]
```diff
@@ -0,0 +1,343 @@
...
+            _stream.WriteLine(gLine);
+        }
+        private void WaitForExtruderTemp()
+        {
+            string gLine = "M109 R" + (decimal)(_machine.ExtruderTemp);
+            _stream.WriteLine(gLine);
+        }
+        private void WaitForCurrentMove()
+        {
+            string gLine = "M400";
```
Comment (henneboy) [inline GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs:334]: why make a string, just pass it as arg

**Review by henneboy — CHANGES_REQUESTED** [review]
Plz fix the comments

File: GOAT-Compiler/Code Generation/CodeGenerator.cs [hunk 20]
```diff
@@ -0,0 +1,961 @@
...
+        {
+            nodeMap.Put(node, -GetValue(node.GetExp()));
+        }
+
+
+        public override void OutAAssignStmt(AAssignStmt node)
+        {
+            Symbol idSymbol = _symbolTable.GetVariableSymbol(node.GetId().Text);
+
+            if (idSymbol != null)
```
[the remark this answers is not in the document]
Comment (henneboy) [reply GOAT-Compiler/Code Generation/CodeGenerator.cs:172]: Same for the other cases

File: GOAT-Compiler/Code Generation/CodeGenerator.cs [hunk 21]
```diff
@@ -0,0 +1,961 @@
...
+            Node left = node.GetL();
+            Node right = node.GetR();
+            if (left != null && right != null)
+            {
+                if (typeMap[left] == Types.Vector && typeMap[right] == Types.Vector)
+                {
+                    Vector Vec1 = ((Vector)GetValue(left));
+                    Vector Vec2 = ((Vector)GetValue(right));
+
+                    Vector resultVec = new Vector(Vec1.X - Vec2.X, Vec1.Y - Vec2.Y, Vec1.Z - Vec2.Z);
```
Comment (henneboy) [inline GOAT-Compiler/Code Generation/CodeGenerator.cs:371]: why not use Vec1 -=Vec2?

File: GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs (hunk 11 above)
Comment (S3JER) [reply GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs:30]: renamed

File: GOAT-Compiler/Code Generation/CodeGenerator.cs [hunk 22]
```diff
@@ -0,0 +1,961 @@
...
+                node.GetTypes().Apply(this);
+            }
+            if (node.GetId() != null)
+            {
+                node.GetId().Apply(this);
+            }
+            {
+                Object[] temp = new Object[node.GetDecl().Count];
+                node.GetDecl().CopyTo(temp, 0);
+                for (int i = 0; i < temp.Length; i++)
```
Comment (henneboy) [inline GOAT-Compiler/Code Generation/CodeGenerator.cs:817]: foreach?

File: GOAT-Compiler/Code Generation/CodeGenerator.cs [hunk 23]
```diff
@@ -0,0 +1,961 @@
...
+            InAProcDecl(node);
+            CallStackRT.Push(new RuntimeTable<Symbol>());
+            if (node.GetId() != null)
+            {
+                node.GetId().Apply(this);
+            }
+            {
+                Object[] temp = new Object[node.GetDecl().Count];
+                node.GetDecl().CopyTo(temp, 0);
+                for (int i = 0; i < temp.Length; i++)
```
Comment (henneboy) [inline GOAT-Compiler/Code Generation/CodeGenerator.cs:846]: foreach?

File: GOAT-Compiler/Code Generation/CodeGenerator.cs [hunk 24]
```diff
@@ -0,0 +1,961 @@
...
+            }
+            if (node.GetBlock() != null)
+            {
+                node.GetBlock().Apply(this);
+            }
+            OutAProcDecl(node);
+        }
+
+
+        public override void InsideScopeOutAProcDecl(AProcDecl node)
```
Comment (henneboy) [inline GOAT-Compiler/Code Generation/CodeGenerator.cs:859]: =>

File: GOAT-Compiler/Code Generation/CodeGenerator.cs [hunk 25]
```diff
@@ -0,0 +1,961 @@
...
+                    case Types.Vector:
+                        return VecMap[key];
+                    default:
+                        return null;
+                }
+            }
+
+            public bool Contains(TKey key)
+            {
+                if (IntMap.ContainsKey(key) ||
```
Comment (henneboy) [inline GOAT-Compiler/Code Generation/CodeGenerator.cs:949]: the if is redundant, just return the expression.
If (bool-exp){
return true;
}
EQUALS:
return bool-exp;

**Review by henneboy — CHANGES_REQUESTED** [review]
Added comments

File: GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs (hunk 13 above)
Comment (S3JER) [reply GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs:113]: Added operator overload on the vector class, "+" and "-"

File: GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs (hunk 14 above)
Comment (S3JER) [reply GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs:117]: Made the helper function

File: GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs (hunk 15 above)
Comment (S3JER) [reply GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs:123]: Added

File: GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs (hunk 17 above)
Comment (S3JER) [reply GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs:262]: added

File: GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs (hunk 16 above)
Comment (S3JER) [reply GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs:273]: added

File: GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs (hunk 18 above)
Comment (S3JER) [reply GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs:306]: added

File: GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs (hunk 19 above)
Comment (S3JER) [reply GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs:334]: fixed

File: GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs (hunk 15 above)
Comment (S3JER) [reply GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs:119]: fixed

File: GOAT-Compiler/Code Generation/CodeGenerator.cs [hunk 26]
```diff
@@ -0,0 +1,824 @@
...
+                {
+                    case Types.Integer:
+                        return IntMap[key];
+                    case Types.FloatingPoint:
+                        return FloatMap[key];
+                    case Types.Boolean:
+                        return BoolMap[key];
+                    case Types.Vector:
+                        return VecMap[key];
+                    default:
```
Comment (henneboy) [inline GOAT-Compiler/Code Generation/CodeGenerator.cs:805]: this cannot happen, so it can be removed

**Review by henneboy — CHANGES_REQUESTED** [review]

File: GOAT-Compiler/Code Generation/CNCMachine.cs (hunk 1 above)
Comment (S3JER) [reply GOAT-Compiler/Code Generation/CNCMachine.cs:30]: fixed

File: GOAT-Compiler/Code Generation/CNCMachine.cs (hunk 2 above)
Comment (S3JER) [reply GOAT-Compiler/Code Generation/CNCMachine.cs:54]: Renamed to HotBedTemp

File: GOAT-Compiler/Code Generation/CNCMachine.cs [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+            }
+        }
+
+        /// <summary>
+        /// The current extruder state.
+        /// </summary>
+        public bool Build { get; set; } = false;
+
+        /// <summary>
+        /// The current rotation of the tutle.
```
[the remark this answers is not in the document]
Comment (S3JER) [reply GOAT-Compiler/Code Generation/CNCMachine.cs:116]: Rewrote

File: GOAT-Compiler/RecSymbolTable.cs (hunk 3 above)
Comment (S3JER) [reply GOAT-Compiler/RecSymbolTable.cs:97]: Fixed

File: GOAT-Compiler/TypeChecker.cs (hunk 9 above)
Comment (S3JER) [reply GOAT-Compiler/TypeChecker.cs:22]: added

File: GOAT-Compiler/TypeChecker.cs (hunk 10 above)
Comment (S3JER) [reply GOAT-Compiler/TypeChecker.cs:445]: Can not do that

File: VisitorTests/CodeGenerator/BuiltInFunctionsTests.cs (hunk 4 above)
Comment (S3JER) [reply VisitorTests/CodeGenerator/BuiltInFunctionsTests.cs:37]: Fixed

File: VisitorTests/CodeGenerator/CodeGeneratorTest.cs (hunk 6 above)
Comment (S3JER) [reply VisitorTests/CodeGenerator/CodeGeneratorTest.cs:26]: fixed

File: GOAT-Compiler/Code Generation/CodeGenerator.cs [hunk 28]
```diff
@@ -0,0 +1,961 @@
...
+            {
+                return RT.Get(symbol, symbol.type);
+            }
+            else
+            {
+                return GlobalRT.Get(symbol, symbol.type);
+            }
+        }
+
+        dynamic GetValue(Node node)
```
[the remark this answers is not in the document]
Comment (S3JER) [reply GOAT-Compiler/Code Generation/CodeGenerator.cs:78]: added

File: GOAT-Compiler/Code Generation/CodeGenerator.cs [hunk 29]
```diff
@@ -0,0 +1,961 @@
...
+        }
+
+
+        public override void CaseADeclProgram(ADeclProgram node)
+        {
+            InADeclProgram(node);
+            {
+                Object[] temp = new Object[node.GetDecl().Count];
+                node.GetDecl().CopyTo(temp, 0);
+                for (int i = 0; i < temp.Length; i++)
```
[the remark this answers is not in the document]
Comment (S3JER) [reply GOAT-Compiler/Code Generation/CodeGenerator.cs]: done

File: GOAT-Compiler/Code Generation/CodeGenerator.cs [hunk 30]
```diff
@@ -0,0 +1,961 @@
...
+            }
+        }
+
+        public override void OutAIdExp(AIdExp node)
+        {
+            Symbol VarSymbol = _symbolTable.GetVariableSymbol(node.GetId().Text);
+            nodeMap.Put(node, GetValue(VarSymbol));
+        }
+
+        public override void OutABoolvalExp(ABoolvalExp node)
```
[the remark this answers is not in the document]
Comment (S3JER) [reply GOAT-Compiler/Code Generation/CodeGenerator.cs]: added

File: GOAT-Compiler/Code Generation/CodeGenerator.cs [hunk 31]
```diff
@@ -0,0 +1,961 @@
...
+            switch (typeMap[node])
+            {
+                case Types.Integer:
+                    nodeMap.Put(node, int.Parse(node.GetNumber().Text));
+                    break;
+                case Types.FloatingPoint:
+                    nodeMap.Put(node, double.Parse(node.GetNumber().Text, CultureInfo.InvariantCulture));
+                    break;
+                default:
+                    throw new Exception("Im litteraly crying right now");
```
[the remark this answers is not in the document]
Comment (S3JER) [reply GOAT-Compiler/Code Generation/CodeGenerator.cs:143]: fixed

File: GOAT-Compiler/Code Generation/CodeGenerator.cs [hunk 32]
```diff
@@ -0,0 +1,961 @@
...
+            }
+        }
+
+
+        public override void OutANegExp(ANegExp node)
```
[the remark this answers is not in the document]
Comment (S3JER) [reply GOAT-Compiler/Code Generation/CodeGenerator.cs]: fixed

File: GOAT-Compiler/Code Generation/CodeGenerator.cs [hunk 33]
```diff
@@ -0,0 +1,961 @@
...
+            BreakFromFunction = true;
+        }
+
+        public override void CaseAStmtlistBlock(AStmtlistBlock node)
+        {
+            InAStmtlistBlock(node);
+            {
+                Object[] temp = new Object[node.GetStmt().Count];
+                node.GetStmt().CopyTo(temp, 0);
+                for (int i = 0; i < temp.Length; i++)
```
[the remark this answers is not in the document]
Comment (S3JER) [reply GOAT-Compiler/Code Generation/CodeGenerator.cs:502]: fixed

File: GOAT-Compiler/Code Generation/CodeGenerator.cs (hunk 25 above)
Comment (S3JER) [reply GOAT-Compiler/Code Generation/CodeGenerator.cs:747]: fixed

File: GOAT-Compiler/Code Generation/CodeGenerator.cs (hunk 26 above)
Comment (S3JER) [reply GOAT-Compiler/Code Generation/CodeGenerator.cs:740]: fixed

**Pull request merged**